### PR TITLE
Bug 1830004: Add listener for keys in search page

### PR DESCRIPTION
**Pull request by bipuladh**
Scrshots:
Screenshot from 2020-04-30 22-57-26
Screenshot from 2020-04-30 22-57-21

Mobile view:
Screenshot from 2020-04-30 21-56-17
/assign @spadgett

### CODE REVIEW

File: frontend/public/components/autocomplete.tsx [hunk 1]
```diff
@@ -83,7 +86,7 @@ const AutocompleteInput: React.FC<AutocompleteInputProps> = (props) => {
           ref={textInputRef}
         />
         <span className="form-control-feedback form-control-feedback--keyboard-hint">
-          <kbd>/</kbd>
+          <kbd>{!isActive ? `/` : 'Esc'}</kbd>
```
[the remark this answers is not in the document]
Comment (bipuladh) [reply frontend/public/components/autocomplete.tsx:89]: Oh okay. I will remove the `Esc`. And yes we were missing on that. I realized while writing this.

File: frontend/public/style/_forms.scss [hunk 2]
```diff
@@ -60,6 +60,10 @@
   }
 }
 
+.co-query-filter:focus + .form-control-feedback--keyboard-hint {
+  right: 10px;
+}
```
[the remark this answers is not in the document]
Comment (bipuladh) [reply frontend/public/style/_forms.scss:65]: yes

File: frontend/public/components/search-filter-dropdown.tsx [hunk 3]
```diff
@@ -1,15 +1,24 @@
...
 
 export enum searchFilterValues {
   Label = 'Label',
   Name = 'Name',
 }
 
+const textInputKeyHandler = {
+  Escape: KeyEventModes.HIDE,
+  '/': KeyEventModes.FOCUS,
+};
```
[the remark this answers is not in the document]
Comment (bipuladh) [reply frontend/public/components/search-filter-dropdown.tsx:14]: I will move it to shared.

File: frontend/packages/console-shared/src/hooks/document-listener.ts [hunk 4]
```diff
@@ -12,12 +12,14 @@ export enum KeyEventModes {
 
 export const useDocumentListener = <T extends HTMLElement>(keyEventMap: KeyEventMap) => {
   const [visible, setVisible] = React.useState(true);
+  const [isActive, setIsActive] = React.useState(false);
```
[the remark this answers is not in the document]
Comment (bipuladh) [reply frontend/packages/console-shared/src/hooks/document-listener.ts:15]: I will remove this since `Esc` is not being used.

File: frontend/public/components/autocomplete.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/components/autocomplete.tsx]: @bipuladh I added the hiding back in #5252

File: frontend/public/components/factory/list-page.jsx [hunk 5]
```diff
@@ -25,66 +25,42 @@ import {
...
+        className={classNames('co-text-filter', className)}
         data-test-id="item-filter"
-        value={value}
-        defaultValue={defaultValue}
-        onChange={onChange}
-        onKeyDown={(e) => e.key === 'Escape' && e.target.blur()}
+        aria-label={placeholder}
         placeholder={placeholder}
-        style={style}
+        onChange={(val) => onChange(val)}
```
Comment (spadgett) [inline frontend/public/components/factory/list-page.jsx:50]: Why not leave it...?

```suggestion
        onChange={onChange}
```

File: frontend/packages/console-shared/src/hooks/document-listener.ts [hunk 6]
```diff
@@ -11,7 +11,14 @@ export enum KeyEventModes {
   FOCUS = 'FOCUS',
 }
 
-export const useDocumentListener = <T extends HTMLElement>(keyEventMap: KeyEventMap) => {
+const textInputKeyHandler = {
+  Escape: KeyEventModes.HIDE,
+  '/': KeyEventModes.FOCUS,
```
Comment (spadgett) [inline frontend/packages/console-shared/src/hooks/document-listener.ts:16]: We should be using the constant defined here:

https://github.com/openshift/console/blob/master/frontend/packages/console-shared/src/constants/common.ts#L16

File: frontend/public/components/RBAC/role.jsx [hunk 7]
```diff
@@ -89,7 +89,7 @@ class Details extends React.Component {
   constructor(props) {
     super(props);
     this.state = {};
-    this.changeFilter = (e) => this.setState({ ruleFilter: e.target.value });
+    this.changeFilter = (e) => this.setState({ ruleFilter: e });
```
Comment (spadgett) [inline frontend/public/components/RBAC/role.jsx:92]: If `e` is not an event, we should give it a different name.

File: frontend/public/components/cluster-settings/global-config.tsx [hunk 8]
```diff
@@ -166,7 +166,7 @@ class GlobalConfigPage_ extends React.Component<GlobalConfigPageProps, GlobalCon
               <TextFilter
                 value={textFilter}
                 label="by name or description"
-                onChange={(e) => this.setState({ textFilter: e.target.value })}
+                onChange={(e) => this.setState({ textFilter: e })}
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/global-config.tsx:169]: If `e` is not an event, we should give it a different name.

```suggestion
                onChange={(textFilter: string) => this.setState({ textFilter })}
```

File: frontend/public/components/events.jsx [hunk 9]
```diff
@@ -209,7 +209,7 @@ export class EventsList extends React.Component {
             <TextFilter
               autoFocus={autoFocus}
               label="Events by name or message"
-              onChange={(e) => this.setState({ textFilter: e.target.value || '' })}
+              onChange={(e) => this.setState({ textFilter: e || '' })}
```
Comment (spadgett) [inline frontend/public/components/events.jsx:212]: If `e` is not an event, we should give it a different name.

File: frontend/public/components/monitoring.tsx [hunk 10]
```diff
@@ -1053,7 +1053,7 @@ const MonitoringListPage = connect(filtersToProps)(
     }
 
     applyTextFilter(e) {
```
Comment (spadgett) [inline frontend/public/components/monitoring.tsx:1055]: If `e` is not an event, we should give it a different name.

File: frontend/public/style/_forms.scss [hunk 11]
```diff
@@ -72,6 +69,10 @@
   padding-right: 25px;
 }
 
+.has-feedback--full-width {
```
Comment (spadgett) [inline frontend/public/style/_forms.scss:72]: Better to prefix console styles with `co-` to avoid name collisions.

File: frontend/public/components/factory/list-page.jsx [hunk 12]
```diff
@@ -25,66 +25,42 @@ import {
...
+        autoFocus={autoFocus}
+        defaultValue={defaultValue}
         tabIndex={0}
+        value={value}
         type="text"
-        autoFocus={autoFocus}
-        aria-label={placeholder}
       />
       <span className="form-control-feedback form-control-feedback--keyboard-hint">
         <kbd>/</kbd>
```
Comment (spadgett) [inline frontend/public/components/factory/list-page.jsx:59]: How do we know the `/` is the shortcut in this context? If `textInputKeyHandler` changes, this could drift from the implementation.

File: frontend/public/components/factory/list-page.jsx (hunk 5 above)
Comment (bipuladh) [reply frontend/public/components/factory/list-page.jsx:50]: `onChange` returns two `(val, event)`. This gives errors(in the console) when passed to a state setter which only takes one argument.

File: frontend/packages/console-shared/src/hooks/document-listener.ts (hunk 6 above)
Comment (bipuladh) [reply frontend/packages/console-shared/src/hooks/document-listener.ts:16]: I had noticed that how about we make document-listener the central place for defining constants related to key bindings? The constant is only used in one other place and that is for namespace dropdown.

File: frontend/public/components/factory/list-page.jsx (hunk 12 above)
Comment (bipuladh) [reply frontend/public/components/factory/list-page.jsx:59]: Are you implying that we should have the same source for defining the key used by the handler and the key shown here? Sorry I am not following you.

Comment (spadgett) [reply frontend/public/components/factory/list-page.jsx:59]: > Are you implying that we should have the same source for defining the key used by the handler and the key shown here? Sorry I am not following you.

Yes. Previously it was this component that added the key handler, but that logic is moved outside of the component now. So this component is hard-coding a hint for a shortcut it doesn't manage.

File: frontend/packages/console-shared/src/hooks/document-listener.ts (hunk 6 above)
Comment (spadgett) [reply frontend/packages/console-shared/src/hooks/document-listener.ts:16]: I'm fine with that, just trying to avoid having it defined in two places.

File: frontend/public/components/factory/list-page.jsx (hunk 5 above)
Comment (spadgett) [reply frontend/public/components/factory/list-page.jsx:50]: Why not pass the event back here as well?

Comment (bipuladh) [reply frontend/public/components/factory/list-page.jsx:50]: > Why not pass the event back here as well?

Do you mean something like `(val, _event) => onChange(val)`

File: frontend/public/components/factory/list-page.jsx [hunk 13]
```diff
@@ -25,66 +25,43 @@ import {
...
         data-test-id="item-filter"
-        value={value}
-        defaultValue={defaultValue}
-        onChange={onChange}
-        onKeyDown={(e) => e.key === 'Escape' && e.target.blur()}
+        aria-label={placeholder}
         placeholder={placeholder}
-        style={style}
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
+        onChange={(val, _event) => onChange(val)}
```
Comment (spadgett) [inline frontend/public/components/factory/list-page.jsx:51]: Sorry, I wasn't clear. I was suggesting

```suggestion
        onChange={onChange}
```

and changing the callback signature to take an event to match the PF `onChange` callback. (It can be ignored by callers who don't care.)

File: frontend/public/style/_forms.scss [hunk 14]
```diff
@@ -53,17 +53,18 @@
...
-  }
+.co-text-filter:focus + .form-control-feedback--keyboard-hint {
+  display: none;
 }
 
 .co-text-filter {
   min-width: 270px;
 }
 
+.co-has-feedback--full-width {
```
Comment (spadgett) [inline frontend/public/style/_forms.scss:64]: This really should be something like `.co-search__text-filter` and be in `_.search.scss`. It's only used there.

File: frontend/public/components/factory/list-page.jsx (hunk 13 above)
Comment (bipuladh) [reply frontend/public/components/factory/list-page.jsx]: Thanks for the clarification. Done!

**Comment by spadgett** [comment]
@bipuladh Looks like you introduced a test failure.

File: frontend/__tests__/components/factory/list-page.spec.tsx [hunk 15]
```diff
@@ -22,11 +23,11 @@ describe(TextFilter.displayName, () => {
   });
 
   it('renders text input', () => {
-    const input: ShallowWrapper<React.InputHTMLAttributes<any>> = wrapper.find('input');
+    const input: ShallowWrapper<any> = wrapper.find(TextInput);
 
     expect(input.props().type).toEqual('text');
     expect(input.props().placeholder).toEqual(`Filter ${label}...`);
-    expect(input.props().onChange).toEqual(onChange);
+    expect(input.props().onChange()).toEqual(((e) => onChange(e))());
```
Comment (spadgett) [inline frontend/__tests__/components/factory/list-page.spec.tsx:30]: Is this change still needed? I thought we went back to `onChange={onChange}`.

Comment (bipuladh) [reply frontend/__tests__/components/factory/list-page.spec.tsx]: Not anymore.

**Comment by spadgett** [comment]
Thanks @bipuladh, looks good 👍
